Sources/Build/BuildDelegate.swift

```diff
@@ -315,4 +328,147 @@ public final class BuildDelegate: BuildSystemDelegate {
+        queue.sync {
+            taskTracker.swiftCompilerDidOuputMessage(message)
+            updateProgress()
+
+            if let output = message.standardOutput {
+                progressAnimation.clear()
+                outputStream <<< output
+                outputStream.flush()
+            }
+        }
+    }
+
+    func swiftCompilerOutputParserDidFail(withError error: Error) {
+        let message = (error as? LocalizedError)?.errorDescription ?? error.localizedDescription
+        diagnostics.emit(data: SwiftCompilerOutputParsingError(message: message))
+        onCommmandFailure?()
+    }
+
+    private func updateProgress() {
+        if let progressText = taskTracker.oldestRunningText {
+            progressAnimation.update(
+                step: taskTracker.finishedCount,
+                total: taskTracker.totalCount,
+                text: progressText)
+        }
+    }
+}
+
+// Tracks tasks based on command status and swift compiler output.
+fileprivate struct TaskTracker {
+    private struct Task {
+        let identifier: String
+        let text: String
+    }
+
+    var isVerbose: Bool = false
+    private var tasks: [Task] = []
+    private(set) var finishedCount = 0
+    private(set) var totalCount = 0
+    private var lastText: String?
+    
+    var oldestRunningText: String? {
+        return tasks.first?.text ?? lastText
+    }
+    
+    mutating func commandStarted(_ command: SPMLLBuild.Command) {
+        let text = isVerbose ? command.verboseDescription : command.description
+        addTask(identifier: command.name, text: text)
+        totalCount += 1
+    }
+    
+    mutating func commandFinished(_ command: SPMLLBuild.Command, result: CommandResult) {
+        removeTask(identifier: command.name)
+
+        switch result {
+        case .succeeded:
+            finishedCount += 1
+        case .cancelled, .failed, .skipped:
+            break
+        }
+    }
+    
+    mutating func swiftCompilerDidOuputMessage(_ message: SwiftCompilerMessage) {
+        switch message.kind {
+        case .began(let info):
+            if let text = message.progressText {
+                addTask(identifier: info.pid.description, text: text)
+            }
+
+            totalCount += 1
+        case .finished(let info):
+            removeTask(identifier: info.pid.description)
+            finishedCount += 1
+        case .signalled(let info):
+            removeTask(identifier: info.pid.description)
+        case .skipped:
+            break
+        }
+    }
+    
+    private mutating func addTask(identifier: String, text: String) {
+        tasks.append(Task(identifier: identifier, text: text))
+    }
+    
+    private mutating func removeTask(identifier: String) {
+        if let index = tasks.index(where: { $0.identifier == identifier }) {
+            if tasks.count == 1 {
+                lastText = tasks[0].text
+            }
+
+            tasks.remove(at: index)
+        }
+    }
+}
```

Comment:
since most methods are mutating, might as well make it a class unless struct is better for some reason?

Reply:
I just have the habit of preferring structs except if I really need reference semantics for some reason (in the case of delegates for example). The Standard Library seems to follow that rule also: for example, `SystemRandomNumberGenerator` only has mutating functions, but it's still a struct. I don't mind either way :-)

Reply:
sounds good, we can make it a class if reference semantics are needed later.